sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/stream/ServerStreamObserverFactory.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.fn.stream;
+
+import io.grpc.stub.CallStreamObserver;
+import io.grpc.stub.StreamObserver;
+
+/**
+ * A {@link StreamObserver} factory that wraps provided {@link StreamObserver}s making them flow
+ * control aware and safe to send data from multiple threads.
+ *
+ * <p>This is different than the {@link StreamObserverFactory} since {@link CallStreamObserver}s can
+ * have their {@link CallStreamObserver#setOnReadyHandler} set directly. This allows us to avoid
+ * needing to use a {@link ForwardingClientResponseObserver} to wrap the inbound {@link
+ * StreamObserver}.
+ */
+public abstract class ServerStreamObserverFactory {
+  /** Wraps the provided outbound {@link StreamObserver}. */
+  public abstract <T> StreamObserver<T> from(StreamObserver<T> outboundObserver);
+
+  /**
+   * Wraps provided outbound {@link StreamObserver}s with a thread-safe and flow control aware
+   * {@link DirectStreamObserver}.
+   */
+  public static ServerStreamObserverFactory direct() {
+    return new Direct();
+  }
+
+  private static class Direct extends ServerStreamObserverFactory {
```

Comment:
Merge this into the StreamObserverFactory class.

You can detect whether its a client stream or server stream by checking whether the object is an instanceof a [ServerCallStreamObserver](https://grpc.io/grpc-java/javadoc/io/grpc/stub/ServerCallStreamObserver.html) or a [ClientCallStreamObserver](https://grpc.io/grpc-java/javadoc/io/grpc/stub/ClientCallStreamObserver.html).

For all the other implementations in StreamObserverFactory, throw an UnsupportedOperationException. Please also update the javadoc mentioning which of the StreamObserverFactory support client or server call stream observers.

Reply:
Done (in the spirit of what you said, but I kept the server-side and client-side factory methods separate intentionally, but at least they have the same API now). This required a fair amount of refactoring. An instanceof check won't work here - in practice the inbound observer may turn out to already be a Beam wrapper over one or the other. I think it's better to keep the intention explicit.